server/Controllers/ApiController.cs

```diff
@@ -146,6 +147,48 @@ public IActionResult PutOne(Guid id, [FromBody]PointJson pointJson)
             return Json(point.ToJson(tags));
         }
 
+        [HttpGet("/api/tags")]
+        public IActionResult GetAllTags()
+        {
+            var tags = _memoryCache.Get<Dictionary<string, Tag>>(_tagsKey);
+            return Json(tags.Values);
+        }
+
+        [HttpPut("/api/tags")]
+        [ValidateModelState]
+        public IActionResult PutAllTags([FromBody]List<Tag> replacementTagsList)
+        {
+            var points = _memoryCache.Get<Dictionary<Guid, Point>>(_pointsKey);
+            var tags = _memoryCache.Get<Dictionary<string, Tag>>(_tagsKey);
+            var duplicateTags = replacementTagsList
+                .Select(tag => tag.Id)
+                .GroupBy(id => id)
+                .Where(g => g.Count() > 1);
+            if (duplicateTags.Any())
+            {
+                return new BadRequestObjectResult(
+                    new InvalidOperationException(
+                        $"Duplicate tag ids: {string.Join(", ", duplicateTags)}").ToString());
+            }
+            var pointsJson = points.Values.Select(point => point.ToJson(tags));
+            var replacementTags = replacementTagsList.ToDictionary(tag => tag.Id, tag => tag);
+            try
+            {
+                pointsJson.Select(p => Point.FromJson(p, replacementTags));
```

Comment:
We do validate the points against tags when reading them from storage. Unifying the tags list processing makes sense, but I wouldn't do it now, as the logic is fairly simple.